### PR TITLE
Typography

### LINKED ISSUE #112: Headings (text)
Add components for headings such as h1, h2, h3, etc. This should include the font-sizes for different breakpoints, margins, font-weight and font-family.

### PR DESCRIPTION
**Pull request by niklasmh**
This builds on #32 and #112. The idea behind this PR is to create a bulletproof typography system that can be used on all content. The main purpose is to make articles and event descriptions look nice, but it should be general enough to be used everywhere else too. The system is greatly inspired by this article: https://uxplanet.org/common-webpage-design-mistakes-59eed9831bd7.

## What has been done:
- Create `<Heading>`, `<Heading2>`, `<Heading3>` (same as `<h1>`, `<h2>`, `<h3>`).
- Create `<Paragraph>` (same as `<p>`)
- Create `<Quote>` (same as `<blockquote>`, thus can provide `by="..."` attribute referring to the source of the quote)
- Create `<Ingress>` (same as italic `<p>`. Spesifically made for articles)
- Create `<Info>` (same as a small `<p>`. Again, spesifically made for articles)
- Create `<Markdown>` component that renders markdown using our components.
- Create a `md` template literal, such that you can parse markdown like ```md`# Title`;```.
- Scale text with screen size. Make it comfortable to read across all devices.

## What may need to be done:
- Add more markdown support (This may belong to another PR as this can be huge and off-topic)

## Current look (also see the chromatic build to review UI changes):
image

## Stories generated:
image

Fixes #112

### CODE REVIEW

**Comment by oleast** [comment]
This looks great to have!

Could we somehow make this work with markdown and have all our text properly formatted?

**Comment by niklasmh** [comment]
@oleast Funny you mentioned it! I am actually working on the markdown at the moment.

I have used a `markdown-to-jsx` library that makes it possible to tweak the ground elements in the markdown parser. It seems to work smooth using headings, quotes, paragraphs and code - which is the main area of this PR. Thus, supporting all kinds of markdown may be relevant in the future.

File: src/components/typography/Text.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  font-size: var(--font-size-md);
+  margin: 2rem 0;
+  ${TextStyle}
+`;
+const StyledQuote = styled.blockquote`
+  font-size: var(--font-size-lg);
+  font-style: italic;
+  text-align: center;
+  margin: 2rem 0;
+  ${TextStyle}
```
Comment (niklasmh) [inline src/components/typography/Text.tsx]: We should probably avoid using too many font styles. Increasing the font and add more spacing should be sufficient enough to make Ingress and Quote "stand-out".

Referring to point 7., 8. and 12. in [0].

[0]: https://uxplanet.org/common-webpage-design-mistakes-59eed9831bd7

Comment (plusk) [reply src/components/typography/Text.tsx]: I think you might be overthinking the nitpickery. When you have a full article written out, most people wouldn't mix too many different styles so close, so it wouldn't be an issue. Ingress is totally fine as is, but we can talk to Peter who does most the article writing to see if he prefers italics or larger font.
For the blockquote, I'd prefer dropping the italics, but that's mostly a personal preferance.
I can look into reviewing the rest some other time

File: src/components/typography/Markdown.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+export const Markdown = ({ children, options = {} }: MarkdownProps): JSX.Element => {
+  return (
+    <MarkdownToJSX
+      options={{
+        overrides: {
+          h1: { component: Heading },
+          h2: { component: Heading2 },
+          h3: { component: Heading3 },
+          p: { component: Paragraph },
+          blockquote: { component: Quote },
```
[the remark this answers is not in the document]
Comment (plusk) [reply src/components/typography/Markdown.tsx]: I don't quite remember where the styling comes from, but we use `react-markdown` on OWF, which does have a list of [node types](https://github.com/rexxars/react-markdown#node-types)

**Comment by niklasmh** [comment]
> When it comes to headings I think it would still be nice to have headings all the way down to `h6`, as we may still need them even though they are not used often.

Another note to that comment is that as markdown supports H1-H6, so we should support that as well. We cannot control what the authors write in markdown, but we can encourage them to stick to H1-H3. H4-H6 is probably most ideal for long texts that really need the hierarchical structure.

> When it comes to their sizes we have to work with changing they sizes in some use-cases because of a11y.

Yup, good point! I haven't looked at that yet, but the articles should be easily readable on the phone.

File: src/components/typography/Heading.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      <Heading>Heading</Heading>
+      <Heading2>Heading2</Heading2>
+      <Heading3>Heading3 (we recommend not going smaller than this)</Heading3>
+    </>
+  </Story>
+</Preview>
+
+## Custom style
+
+The component behaves just like a normal heading tag. Just use the `style` attribute if you need modifications.
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/typography/Heading.stories.mdx]: That assumes they are using `styled-components` as a part of their system, though that would work for OWF.

File: src/components/typography/Heading.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  return <StyledH1 {...props}>{children}</StyledH1>;
+};
+
+export const Heading2 = ({ children, ...props }: HeaderProps): JSX.Element => {
+  return <StyledH2 {...props}>{children}</StyledH2>;
+};
+
+export const Heading3 = ({ children, ...props }: HeaderProps): JSX.Element => {
+  return <StyledH3 {...props}>{children}</StyledH3>;
+};
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/typography/Heading.tsx]: Not a bad idea!

File: src/components/typography/Markdown.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/typography/Markdown.tsx]: Here is the complete list of tags that are used by the library: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/markdown-to-jsx/index.d.ts#L31-L66

What I have not thought of yet is:
- Tables
- Images
- Code
- Lists
- Horizontal lines
- Sub and sup

I may be smart to replace each of these as we do not want to depend on global styles. At least make it hard to accidentally style the markdown content globally.

**Comment by niklasmh** [comment]
By using chromatic we can now see that my changes affected two other stories: Tabs, Media Query Breakpoints. Just see the details in the chromatic build check.

File: src/components/typography/Text.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  color: ${colors.grayDarken60};
+  float: right;
+  font-size: var(--font-size-md);
+
+  &::before {
+    content: '- ';
+  }
+`;
+
+export const Quote: FC<QuoteProps> = ({ children, by = '', ...props }: QuoteProps): JSX.Element => {
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/typography/Text.tsx]: Agree. Looking at how kode24 does it, just plain quotes make it nice.

File: src/common/global.ts [hunk 6]
```diff
@@ -1,24 +1,51 @@
...
+      }px) / ${breakPointsInPX.minDesktopWidth - breakPointsInPX.minTabletWidth})));
+    }
+    @media ${media.desktopAndUp} {
+      --font-size: calc(var(--font-base-size) * var(--font-max-scale));
+    }
   }
-  h1 {
-    font-size: 2.5rem;
+  h1, h2, h3, h4, h5, h6 {
+    font-family: 'Source Code Pro', monospace;
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/common/global.ts]: I will just change it back to `'Source Sans Pro', sans-serif`, as we use that on OWF. If someone feels for something else it should be easy to fix.

File: src/components/typography/Heading.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  line-height: 3.125rem;
+  ${HeadingStyle}
+`;
+const Heading4 = styled.h4`
+  font-size: calc(var(--font-size-xxl) * 0.6);
+  line-height: 2.5rem;
+  ${HeadingStyle}
+`;
+const Heading5 = styled.h5`
+  font-size: calc(var(--font-size-xxl) * 0.53);
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/typography/Heading.tsx]: I have no idea how to fix font sizes, thus just using the xxl, xl and lg for h1-h3 should be sufficient enough. The h4-h6 becomes smaller than normal text, but adding boldness to them makes them stick out as headers, so I will just leave it there.

File: src/components/typography/Heading.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import styled, { css } from 'styled-components';
+import { colors } from 'common/colors';
+
+const HeadingStyle = css`
+  color: ${colors.grayDarken90};
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/typography/Heading.tsx]: Seems like a good idea to have that as a default.

File: src/components/icon/Icon.tsx [hunk 9]
```diff
@@ -3,16 +3,21 @@ import styled, { StyledComponentProps } from 'styled-components';
...
   font-size: inherit;
-  vertical-align: bottom;
+  vertical-align: middle;
 `;
 
-const Icon = ({ name }: IconProps) => {
-  return <StyledIcon className="material-icons-outlined">{name}</StyledIcon>;
+const Icon = ({ name, className = '', ...props }: IconProps) => {
+  return (
+    <StyledIcon className={`material-icons-outlined ${className}`} {...props}>
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/icon/Icon.tsx]: I think I added this such that Icon could be extended the styled-components way of doing things. There could be other ways of adding a classname, but I have not found any yet so I think this is one of the safest approaches so far.

File: src/components/icon/Icon.tsx [hunk 10]
```diff
@@ -3,16 +3,21 @@ import styled, { StyledComponentProps } from 'styled-components';
...
 interface IconProps extends StyledComponentProps<'i', any, any, any> {
   name: string;
+  className?: string;
 }
 
 const StyledIcon = styled.i`
   color: inherit;
   font-size: inherit;
-  vertical-align: bottom;
+  vertical-align: middle;
```
[the remark this answers is not in the document]
Comment (niklasmh) [reply src/components/icon/Icon.tsx]: I am actually not sure how Icons "should" align, but from our storybook example, the `middle` seems to work best:

This is with `baseline`:
image

This is with `middle`:
image

This is with `bottom` (which was there previously):
image

**Pull request merged**